### PR TITLE
Grinder improvements / additions

### PR DESCRIPTION
**Pull request by Starkku**
You can now customize which types of objects a building with `Grinding` set can grind as well as the grinding sound.
- `Grinding.AllowAllies` changes whether or not to allow units to enter allies' buildings.
- `Grinding.AllowOwner` changes whether or not to allow units to enter your own buildings.
- `Grinding.AllowTypes` can be used to define InfantryTypes and VehicleTypes that can be grinded by the building. Listing any will disable grinding for all types except those listed.
- `Grinding.DisallowTypes` can be used to exclude InfantryTypes or VehicleTypes from being able to enter the grinder building.
- `Grinding.Sound` is a sound played by when object is grinded by the building. If not set, defaults to `[AudioVisual]`->`EnterGrinderSound`.
- `Grinding.Weapon` is a weapon fired at the building & by the building when it grinds an object. Will only be fired if at least weapon's `ROF` amount of frames have passed since it was last fired. Note: for this to work it requires that the WeaponType used has been initialized prior to parsing of this key, from a previously parsed TechnoType or from `[WeaponTypes]`, which is only available with Ares.

In `rulesmd.ini`:
```ini
[SOMEBUILDING]             ; BuildingType
Grinding.AllowAllies=false ; boolean
Grinding.AllowOwner=true   ; boolean
Grinding.AllowTypes=       ; List of InfantryTypes / VehicleTypes
Grinding.DisallowTypes=    ; List of InfantryTypes / VehicleTypes
Grinding.Sound=            ; Sound
Grinding.Weapon=           ; WeaponType
```

In addition following minor improvements / bugfixes are included:
- Fixed interaction of `UnitAbsorb` & `InfantryAbsorb` with `Grinding` buildings. The keys will now make the building only accept appropriate types of objects.
- Fixed missing 'no enter' cursor for VehicleTypes being unable to enter a `Grinding` building.
- Fixed Engineers being able to enter `Grinding` buildings even when they shouldn't (such as ally building at full HP).

Should work as expected with the AI script action 'Enter grinder' and the Ares bugfix for Ivan & Grinder interaction.

### CODE REVIEW

**Comment by brsajo** [comment]
In my testing:

- `Grinding.AllowAllies=true` works as expected.
- `Grinding.AllowedTypes=INIT` doesn't seem to have any effect on what types are allowed to enter.
- `Grinding.DisallowedTypes=LTNK` doesn't seem to have any effect on what types are allowed to enter.
- `Grinding.Sound=CloningVatsCreate` works as expected.
- `UnitAbsorb` & `InfantryAbsorb` fixes confirmed working as expected.
- 'no enter' cursor for VehicleTypes fix confirmed working as expected.
- Engineers entering grinders fix confirmed working as expected.
Haven't tested the 'Enter grinder' AI script.

**Comment by CommanderNour** [comment]
Tested every tag works as intended even the Allow & Disallow Types, Good job man
& sorry for not providing any screen shots

**Comment by SMxReaver** [comment]
grindweapon

So far, testing: 
Grinding.AllowAllies=false stops you from entering an allies' grinder but not your own, would we need a hypothetical Grinding.AllowOwner for this functionality?
Grinding.AllowTypes, Grinding.DisallowTypes, Grinding.Weapon, Grinding.Sound all working as intended

UnitAbsorb and InfantryAbsorb fixes seem to be working correctly too. Have not tested Engineer or script actions at this moment.

**Comment by Starkku** [comment]
Fixed `Grinding.Weapon` generating INI parsing errors and added `Grinding.AllowOwner`.

File: Phobos.vcxproj [hunk 1]
```diff
@@ -75,6 +75,7 @@
     <ClCompile Include="src\Ext\Unit\Hooks.DisallowMoving.cpp" />
     <ClCompile Include="src\Ext\WarheadType\Detonate.cpp" />
     <ClCompile Include="src\Ext\Techno\Hooks.Shield.cpp" />
+	<ClCompile Include="src\Ext\Techno\Hooks.Grinding.cpp" />
```
Comment (Metadorius) [inline Phobos.vcxproj:78]:
```suggestion
    <ClCompile Include="src\Ext\Techno\Hooks.Grinding.cpp" />
```

File: src/Ext/Techno/Hooks.Grinding.cpp [hunk 2]
```diff
@@ -0,0 +1,188 @@
+#include "Body.h"
+
+#include <InfantryClass.h>
+
+#include <Ext/Building/Body.h>
+#include <Ext/BuildingType/Body.h>
+
+DEFINE_HOOK(0x43C30A, BuildingClass_ReceiveMessage_Grinding, 0x6)
+{
+	enum { Return_Static = 0x43C31A, Return_Negative = 0x43CB68, Return_Roger = 0x43CCF2 };
```
Comment (Metadorius) [inline src/Ext/Techno/Hooks.Grinding.cpp:10]: Let's keep it strictly `PascalCase`, not `Some_Mixed_Case`.

**Review by Metadorius — CHANGES_REQUESTED** [review]

File: src/Ext/Techno/Hooks.Grinding.cpp (hunk 2 above)
Comment (Starkku) [reply src/Ext/Techno/Hooks.Grinding.cpp:10]: Pascal sliced into bits.

Comment (Metadorius) [reply src/Ext/Techno/Hooks.Grinding.cpp]: Too delicious.

**Pull request merged**